### PR TITLE
Avoid some closure allocations in ilxgen hot paths

### PR DESCRIPTION
**Pull request by jack-pappas**
After running some profiling on the compiler, I noticed that it was allocating closures in some hot paths within the ``ilxgen`` phase. I haven't fixed every single one, but I did fix some of the simple cases where calls to ``List.iter``, ``FlatList.iter``, etc. could be easily replaced by an imperative-style ``for`` loop.

### CODE REVIEW

**Comment by latkin** [comment]
Can you share your analysis logs?

`for` loops over lists are done via `IEnumerable`, so there is still allocation. Have you measured it to be noticeably cheaper?

File: src/fsharp/ilxgen.fs [hunk 1]
```diff
@@ -5662,8 +5665,9 @@ and AllocTopValWithinExpr cenv cgbuf cloc scopeMarks v eenv =
 and EmitSaveStack cenv cgbuf eenv m scopeMarks =
     let savedStack = (cgbuf.GetCurrentStack())
     let savedStackLocals,eenvinner = List.mapFold (fun eenv ty -> AllocLocal cenv cgbuf eenv true (ilxgenGlobalNng.FreshCompilerGeneratedName ("spill",m), ty) scopeMarks) eenv savedStack
-    List.iter (EmitSetLocal cgbuf) savedStackLocals;
-    cgbuf.AssertEmptyStack();
+    for ssl in savedStackLocals do
```
Comment (dsyme) [inline src/fsharp/ilxgen.fs]: Can you change or inline "List.iter" (including in local.fs) to get the same perfomance effect?  That would help all F# code without requiring these changes.

File: src/fsharp/ilxgen.fs [hunk 2]
```diff
@@ -70,7 +70,7 @@ let mkLdfldMethodDef (ilMethName,reprAccess,isStatic,ilTy,ilFieldName,ilPropType
    ilMethodDef |> AddSpecialNameFlag
 
 let ChooseParamNames fieldNamesAndTypes = 
-    let takenFieldNames = fieldNamesAndTypes |> List.map p23 |> Set.ofList
+    let takenFieldNames = fieldNamesAndTypes |> Seq.map p23 |> Set.ofSeq
```
Comment (dsyme) [inline src/fsharp/ilxgen.fs]: It may be better to avoid "Set" here at all.

**Comment by dsyme** [comment]
Looking at this, I do think simply inlining List.iter may give the desired effect, and be a good change for everyone.  But either way, a perf analysis needs to be done.

I really, really like small changes that improve compiler perf bit by bit. It's great and it's what we need to do.  However I do also wonder if we should get consistent, reliable CI-based compiler perf testing in place with a checkin-by-checkin graph of compiler performance sooner rather than later.  If we improve perf, we need to be able to see it in the numbers, and it is much easier to see it and believe it in an automated system, rather than having @latkin and others verify the perf improvements by hand.

@forki is working on some CI at the moment.  If we get a dedicated stable CI machine (rather than VMs) then we can probably set up a regular compiler perf analysis as part of the CI build, though we'd still have to make sure that it was a very stable and reliable machine configuration.

**Comment by isaacabraham** [comment]
That's a good idea, but with many PAAS / SAAS services it can be difficult to get guaranteed performance for short time frames, although I do think that AppVeyor does offer something like a premium service, but again I'm not certain if its guaranteed to have equal performance across sessions.

**Comment by latkin** [comment]
Thank you for the submission. Per the [contribution guidelines](https://github.com/Microsoft/visualfsharp/blob/fsharp4/CONTRIBUTING.md#mimimum-bar-for-performance-improvement-pull-requests), perf optimizations should include tests and quantitative analysis showing reliable gains from the change, which others can reproduce.  Please re-open when tests and analysis are in place.

**Pull request closed without merging**